src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala

```diff
@@ -13,34 +13,377 @@ import spray.http.StatusCodes
+
+    def entityRef = foreignKey("FK_ENT_ATTRIBUTE_ENTITY_REF", valueEntityRef, entityQuery)(_.id.?)
+    def parentEntity = foreignKey("FK_ATTRIBUTE_PARENT_ENTITY", entityId, entityQuery)(_.id)
+  }
+
+  class WorkspaceAttributeTable(tag: Tag) extends Table[WorkspaceAttributeRecord](tag, "WORKSPACE_ATTRIBUTE") {
+    def id = column[Long]("id", O.PrimaryKey, O.AutoInc)
+    def workspaceId = column[UUID]("workspace_id")
     def name = column[String]("name")
     def valueString = column[Option[String]]("value_string")
     def valueNumber = column[Option[Double]]("value_number")
     def valueBoolean = column[Option[Boolean]]("value_boolean")
-    def valueEntityRef = column[Option[UUID]]("value_entity_ref")
+    def valueEntityRef = column[Option[Long]]("value_entity_ref")
     def listIndex = column[Option[Int]]("list_index")
 
-    def * = (id, name, valueString, valueNumber, valueBoolean, valueEntityRef, listIndex) <> (AttributeRecord.tupled, AttributeRecord.unapply)
+    def * = (id, workspaceId, name, valueString, valueNumber, valueBoolean, valueEntityRef, listIndex) <> (WorkspaceAttributeRecord.tupled, WorkspaceAttributeRecord.unapply)
+
+    def entityRef = foreignKey("FK_WS_ATTRIBUTE_ENTITY_REF", valueEntityRef, entityQuery)(_.id.?)
+    def workspace = foreignKey("FK_ATTRIBUTE_PARENT_WORKSPACE", workspaceId, workspaceQuery)(_.id)
+  }
+
+  class SubmissionAttributeTable(tag: Tag) extends Table[SubmissionAttributeRecord](tag, "SUBMISSION_ATTRIBUTE") {
+    def id = column[Long]("id", O.PrimaryKey, O.AutoInc)
+    def workspaceId = column[UUID]("workspace_id")
+    def name = column[String]("name")
+    def valueString = column[Option[String]]("value_string")
+    def valueNumber = column[Option[Double]]("value_number")
+    def valueBoolean = column[Option[Boolean]]("value_boolean")
+    def valueEntityRef = column[Option[Long]]("value_entity_ref")
+    def listIndex = column[Option[Int]]("list_index")
+
+    def * = (id, name, valueString, valueNumber, valueBoolean, valueEntityRef, listIndex) <> (SubmissionAttributeRecord.tupled, SubmissionAttributeRecord.unapply)
+
+    def entityRef = foreignKey("FK_SUB_ATTRIBUTE_ENTITY_REF", valueEntityRef, entityQuery)(_.id.?)
+  }
+
+  protected object entityAttributeQuery extends TableQuery(new EntityAttributeTable(_)) {
+
+    /**
+     * Insert an attribute into the database. This will be multiple inserts for a list and will lookup
+     * referenced entities.
+     *
+     * @param name
+     * @param attribute
+     * @param workspaceId used only for AttributeEntityReferences (or lists of them) to resolve the reference within the workspace
+     * @return a sequence of write actions the resulting value being the attribute id inserted
+     */
+    def insertAttributeRecords(entityId: Long, name: String, attribute: Attribute, workspaceId: UUID): Seq[ReadWriteAction[Int]] = {
```

Comment:
The similarity between these methods and those similarly named below worries me a great deal from a code maintenance perspective. Can you factor out the common functionality? I can offer some suggestions if that would help.